avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java

```diff
@@ -60,44 +76,63 @@ public PrimitiveFloatList(Collection<Float> c) {
    * @throws IOException on io errors
    */
   public static Object readPrimitiveFloatArray(Object old, Decoder in) throws IOException {
-    long l = in.readArrayStart();
-    if (l > 0) {
-      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old, (int) l);
+    long length = in.readArrayStart();
+    long totalLength = 0;
+
+    if (length > 0) {
+      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old);
+
       do {
-        for (long i = 0; i < l; i++) {
-          array.addPrimitive(in.readFloat());
-        }
-        l = in.arrayNext();
-      } while (l > 0);
+        // Allocate ByeBuffer of size array length * Float.BYTES to hold the array values
+        ByteBuffer byteBuffer = ByteBuffer.allocate((int)length * FLOAT_SIZE).order(ByteOrder.LITTLE_ENDIAN);
+        in.readFixed(byteBuffer.array(), 0, (int)(length * FLOAT_SIZE));
+        array.byteBuffers.add(byteBuffer);
+        totalLength += length;
+        length = in.arrayNext();
+      } while (length > 0);
+
+      setupElements(old, array, (int)totalLength);
       return array;
     } else {
-      return newPrimitiveFloatArray(old, 0);
+      return new PrimitiveFloatList(0);
     }
   }
 
+  private static void setupElements(Object old, PrimitiveFloatList list, int size) {
+    if (old instanceof PrimitiveFloatList) {
+      PrimitiveFloatList oldFloatList = (PrimitiveFloatList) old;
+      if (size <= oldFloatList.getCapacity()) {
+        // reuse the float array directly
+        oldFloatList.clear();
+      } else {
+        oldFloatList.resizeAndClear(size);
+      }
+      return;
+    }
+    list.elements = new float[size];
+    list.size = size;
+  }
+
   /**
-   * @param expected {@link Schema} to inspect
-   * @return true if the {@code expected} SCHEMA is of the right type to decode as a {@link PrimitiveFloatList}
-   *         false otherwise
-   */
+     * @param expected {@link Schema} to inspect
+     * @return true if the {@code expected} SCHEMA is of the right type to decode as a {@link PrimitiveFloatList}
+     *         false otherwise
+     */
   public static boolean isFloatArray(Schema expected) {
     return expected != null && Schema.Type.ARRAY.equals(expected.getType()) && FLOAT_SCHEMA.equals(
         expected.getElementType());
   }
 
-  private static Object newPrimitiveFloatArray(Object old, int size) {
+  private static Object newPrimitiveFloatArray(Object old) {
     if (old instanceof PrimitiveFloatList) {
```

Comment:
Should `isCached` be flipped back to false when re-using an old instance? Also, this should probably happen inside a synchronized block? Let's think about how to reduce the critical section to the smallest possible number of lines...

Reply:
Done